### ISSUE TITLE
windows build error

### ISSUE THREAD

**Issue by yangmmm** [AdaptiveParticles/LibAPR#88]
I met a link error when I was trying to build LibAPR.
My system is windows 10
CMake version is 3.11.1
Compiler is  Intel 18.0.2.20180210
when I ran
```
cmake -G "Visual Studio 15 2017 Win64" -DTIFF_INCLUDE_DIR="C:/Program Files/tiff/include" -DTIFF_LIBRARY="C:/Program Files/tiff/lib/tiff.lib " -DHDF5_ROOT="C:/Program Files/HDF_Group/HDF5/1.10.2" -DAPR_PREFER_EXTERNAL_BLOSC="OFF" -T "Intel C++ Compiler 18.0" ..
```
I got
```-- The C compiler identification is Intel 18.0.2.20180210
-- The CXX compiler identification is Intel 18.0.2.20180210
-- Check for working C compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2018/windows/bin/intel64/icl.exe
-- Check for working C compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2018/windows/bin/intel64/icl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2018/windows/bin/intel64/icl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/IntelSWTools/compilers_and_libraries_2018/windows/bin/intel64/icl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Configuring for APR version: 1.0.1
-- Found HDF5: C:/Program Files/HDF_Group/HDF5/1.10.2/lib/hdf5.lib (found version "1.10.2")
-- Found ZLIB: C:/Users/Server 36/Anaconda3/Library/lib/z.lib (found version "1.2.8")
-- Found TIFF: C:/Program Files/tiff/lib/tiff.lib (found version "4.0.9")
-- Found OpenMP_C: -Qopenmp (found version "5.0")
-- Found OpenMP_CXX: -Qopenmp (found version "5.0")
-- Found OpenMP: TRUE (found version "5.0")
-- APR: blosc not found, using internal blosc
Configuring for Blosc version: 1.11.2
-- Using LZ4 internal sources.
-- Using Snappy internal sources.
-- No Zstd library found.  Using internal sources.
-- Building for system processor AMD64
-- Adding run-time support for SSE2
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- APR: Install library in [C:/Program Files/LibAPR]
-- APR: Building examples
-- Configuring done
-- Generating done
-- Build files have been written to: E:/LibAPR/build
```

And then I ran `cmake --build . --config Debug`
Then the error came up:
```
  blosclz.c
  shuffle-generic.c
  bitshuffle-generic.c
  shuffle-sse2.c
  bitshuffle-sse2.c
  lz4.c
E:\LibAPR\external\c-blosc\internal-complibs\lz4-1.7.5\lz4.c(184): error : expected a type specifier [E:\LibAPR\build\e
xternal\c-blosc\blosc\blosc_static.vcxproj]
    typedef union { U16 u16; U32 u32; reg_t uArch; } __attribute__((packed)) unalign;
                                                                   ^

E:\LibAPR\external\c-blosc\internal-complibs\lz4-1.7.5\lz4.c(184): error : expected a ";" [E:\LibAPR\build\external\c-b
losc\blosc\blosc_static.vcxproj]
    typedef union { U16 u16; U32 u32; reg_t uArch; } __attribute__((packed)) unalign;
                                                                             ^

E:\LibAPR\external\c-blosc\internal-complibs\lz4-1.7.5\lz4.c(186): error : identifier "unalign" is undefined [E:\LibAPR
\build\external\c-blosc\blosc\blosc_static.vcxproj]`
And also error as followed:
`CustomBuild:
  Building Custom Rule E:/LibAPR/CMakeLists.txt
  CMake does not need to re-run because E:/LibAPR/build/CMakeFiles/generate.stamp is up-to-date.
Lib:
  C:\Program Files (x86)\IntelSWTools\compilers_and_libraries_2018\windows\bin\Intel64\xilib.exe -qm64 /OUT:"E:\LibAPR\
  build\Debug\apr.lib" /NOLOGO /MACHINE:X64 E:\LibAPR\build\aprObjLib.dir\Debug\blosc_filter.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\hdf5functions_blosc.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\APRRaycaster.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\Camera.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\Object.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\RaytracedObject.obj  /machine:x64 ""$"<TARGET_FILE:blosc_static>"
LINK : : error LNK1104: can't open file“$<TARGET_FILE:blosc_static>” [E:\LibAPR\build\staticLib.vcxproj]
```

It looks like a blosc link error, but I don't know how to fix it. I tried to compile the blosc library in https://github.com/Blosc/c-blosc and add a '-DAPR_PREFER_EXTERNAL_BL
OSC="ON"' in the CMake commond, but it didn't work.
Would you please help me？

Edit by @skalarproduktraum: Improved formatting for better readability

**Comment by cheesema** [comment]
Hi yangmmm,

Thanks for letting us know. Will get back to you shortly hopefully with a fix.

Cheers,
B

**Comment by skalarproduktraum** [comment]
Hello @yangmmm,

the issue you're seeing is caused by the LZ4 library not correctly detecting the Intel Compiler on Windows, which does not support the __attribute__((packed)) notation, as it's Linux cousin does.

There are three options to solve this:
* rerun the first cmake command with `-DLZ4_FORCE_MEMORY_ACCESS=1` and rebuild (I haven't tried that one, but that's what the header file is suggesting)
* remove the `__attribute((packed))__` part from the declaration
* force CMake to use your externally installed blosc library, although I do not have an idea how to point it to the installation directory on Windows

Please let us know if any of these solved your problem 👍

**Comment by yangmmm** [comment]
Thanks a lot @skalarproduktraum and @cheesema ,
I have tried the first solution, by adding `-DLZ4_FORCE_MEMORY_ACCESS=2`, but the CMake didn't recognize and return an warming `CMake Warning: Manually-specified variables were not used by the project: LZ4_FORCE_MEMORY_ACCESS` .
And I deleted the `__attribute((packed))__` declaration in lz4.c file, and it worked! But some new error came out:

> E:\LibAPR\build\aprObjLib.dir\Debug\hdf5functions_blosc.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\APRRaycaster.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\Camera.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\Object.obj
  E:\LibAPR\build\aprObjLib.dir\Debug\RaytracedObject.obj  /machine:x64
LINK : : warning LNK4044: Unrecognized options“/Wl,--whole-archive”；Ignored [E:\LibAPR\build\sharedLib.vcxproj]
LINK : : warning LNK4044: Unrecognized options“/Wl,--no-whole-archive”；Ignored [E:\LibAPR\build\sharedLib.vcxproj]
    Building libray E:/LibAPR/build/Debug/apr.lib and object E:/LibAPR/build/Debug/apr.exp
blosc_filter.obj : : error LNK2019:Unresolved external symbols H5E_CANTREGISTER_g，The symbol is referenced in function register_blosc  [E:\LibAPR\build\sharedLib.
vcxproj]
blosc_filter.obj : : error LNK2019:Unresolved external symbols H5E_PLINE_g，The symbol is referenced in function register_blosc  [E:\LibAPR\build\sharedLib.vcxproj
]
blosc_filter.obj : : error LNK2019:Unresolved external symbols H5E_ERR_CLS_g，The symbol is referenced in function register_blosc  [E:\LibAPR\build\sharedLib.vcxpr
oj]
blosc_filter.obj : : error LNK2019:Unresolved external symbols H5E_CALLBACK_g，The symbol is referenced in function blosc_set_local  [E:\LibAPR\build\sharedLib.vcx
proj]
hdf5functions_blosc.obj : : error LNK2019:Unresolved external symbols H5P_CLS_DATASET_CREATE_ID_g，The symbol is referenced in function "void __cdecl hdf5_write_data_b
losc(__int64,__int64,char const *,unsigned __int64,unsigned __int64 *,void *,unsigned int,unsigned int,unsigned int)" (
?hdf5_write_data_blosc@@YAX_J0PEBD_KPEA_KPEAXIII@Z)  [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_UINT64_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_FLOAT_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_INT_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_INT16_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_UINT8_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_C_S1_g [E:\LibAPR\build\sharedLib.vcxproj]
APRRaycaster.obj : : error LNK2001:Unresolved external symbols H5T_NATIVE_UINT16_g [E:\LibAPR\build\sharedLib.vcxproj]
libblosc.lib(blosc.obj) : : error LNK2019:Unresolved external symbols compress2，The symbol is referenced in function zlib_wrap_compress  [E:\LibAPR\build\sharedLi
b.vcxproj]
libblosc.lib(blosc.obj) : : error LNK2019:Unresolved external symbols uncompress，The symbol is referenced in function zlib_wrap_decompress  [E:\LibAPR\build\share
dLib.vcxproj]

I also tried to build the project by adding  '-DAPR_PREFER_EXTERNAL_BLOSC="ON" , in order to force CMake to use  the externally installed blosc library. Unfortunately, it have the same result as  deletion of `__attribute((packed))__`.
Could you give me some advice?  
I am very confuse now, maybe I should take a C++ lesson?

**Comment by cheesema** [comment]
@skalarproduktraum any ideas?

**Comment by cheesema** [comment]
This is still broken. 

We are working on this and hopefully will have it resolved soon.

**Comment by MSusik** [comment]
@cheesema 
I've never been an advanced Windows user, but I heard it is possible to run a Linux Docker container on Windows. If the users want to use LibAPR on Windows, they can use the Dockerfile provided, until this issue is solved. Hope that helps :)

**Comment by krzysg** [comment]
Unless you provide easy way to use LibAPR via Docker in Fiji/ImageJ (including automatic installation) it is not solution we are looking for.

**Comment by cheesema** [comment]
Recent refactoring of the library may remove these issue. Fingers crossed, we will see soon.

**Comment by cheesema** [comment]
It seems there is progress here. Currently mingw build is 90% there, will need to also check the Visual Studio + Intel Compiler route.

**Comment by psteinb** [comment]
According to [this](https://gitlab.kitware.com/cmake/cmake/issues/19404#note_621517) entry in my issue with cmake, you may be
able to compile OpenMP code with LLVM under Windows with LLVM 9.0.0. Final Release was scheduled for August 28. So there is a
high likelihood, that it will be available soon. I'll add a note there that the community hope, cmake will be aware of this.

**Comment by cheesema** [comment]
Thanks @psteinb I actually have been playing round with this to understand the issues and had converged exactly to your post ;) Fingers crossed.. seems so close, yet so far. Given that the command line version can work, is there a possible work around?

**Comment by psteinb** [comment]
Well, effectively one would have to write a custom cmake module to identify/pass the clang flags in the correct fashion on
windows for that matter. As soon as llvm/clang-cl has fixed the problem under windows, this cmake module will become unused. But
at least the project can move forward until then. ;)

**Comment by cheesema** [comment]
The current easiest solution in windows is to utilise WSL2, here installation runs fine. One signficant limitation is the current access speed between the windows and linux file subsystem. However, I believe this should be fixed this year (wishful thinking)

**Comment by cheesema** [comment]
#140 some success here WIP.

**Comment by cheesema** [comment]
This should now be resolved with instructions in the Readme from #140
